### PR TITLE
[WIP] Teach MCO to use the new-format image by default 

### PR DESCRIPTION
**Pull request by jkyros**
#### This is a way to give the MCO "new format" image support by: 
- Plumbing `BaseOperatingSystemContainer` and `BaseOperatingSystemExtensionsContainer` all the way through into MachineConfig like OSImageURL

I don't think this is actually how we'll want to do it, this is just the "simplest way" that doesn't "overload" `machine-os-content` with a new format image. 

#### Caveats: 
- *This should not merge*. The moment that https://github.com/openshift/machine-config-operator/commit/a83a3f96cc50f8c68a27e948608ba1b6bebe05ec goes in, the new format images will become required and break everything
- This includes Colin's https://github.com/openshift/machine-config-operator/pull/3017, since we haven't merged it yet, but I would obviously take it out when it does merge. 
- If updating into a release using this, this will work as long as `rhel-coreos-8` (and optionally `rhel-coreos-8-extensions`)  are included in the release
- If building a cluster, it will fail unless you're using an installer that supplies the image-references to the MCO (something like this: https://github.com/jkyros/installer/blob/allow-mco-consume-multiple-images/data/data/bootstrap/files/usr/local/bin/bootkube.sh.template#L2977 ) 

#### Can I install a cluster using the new format image? Yes, but right now you probably don't want to, because you have to:  

1. Build an installer, say off this branch https://github.com/jkyros/installer/tree/allow-mco-consume-multiple-images
2. Create a payload manually
```
oc adm release new -n ocp --server https://api.ci.l2s4.p1.openshiftapps.com --from-release registry.ci.openshift.org/ocp/releas
e:4.12 --to-image quay.io/jkyros/ocp-release:v4.12  --include=rhel-coreos-8-extensions rhel-coreos-8-extensions=registry.ci.openshift.org/rhcos-dev
el/rhel-coreos-extensions:latest --include=rhel-coreos-8 rhel-coreos-8=quay.io/jkyros/derived-images:rhel-coreos-latest  machine-config-operator=quay.io/jkyros/ma
chine-config-operator:latest 
```
4. Install using your custom installer, overriding the payload: 
 ```
export OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE=quay.io/jkyros/ocp-release:v4.12
../openshift-install-dev create cluster ...
```

This PR is in service to: https://issues.redhat.com/browse/MCO-291

### CODE REVIEW

**Comment by cgwalters** [comment]
OK!  Word on the street is that this *may* be unblocked.  Basically I think we can now refer to `rhel-coreos-8` in the MCO, and that should cause it to actually appear in the release image.

**Comment by cgwalters** [comment]
Just going to lift draft and see what happens in CI...

**Comment by cgwalters** [comment]
/retest
build02 was ill

**Comment by cgwalters** [comment]
randomly trying https://github.com/openshift/release/pull/31083

**Comment by cgwalters** [comment]
@jkyros did you do a PR for openshift/installer to have it pass through the new image references?  We'll need that in order for bootstrap to work, right?

**Comment by cgwalters** [comment]
I believe we need to land https://github.com/openshift/machine-config-operator/pull/3258/commits/c63533d340f600403a28003f4e539b14315f844c first, then PR to the installer

**Comment by cgwalters** [comment]
@yuqi-zhang and I were talking about this and we think it makes sense to split this PR in two:

- code changes sufficient to just reference the new image, but *not use it*
- a second PR which actually tries to use it

The first hunk of this could land right now with no external dependencies (e.g. on the installer etc.)

**Comment by yuqi-zhang** [comment]
We discussed this last week and I think we have reached consensus on these steps or so:
1. get https://github.com/openshift/machine-config-operator/pull/3272 in
2. split out this PR to just consume to the controllerconfig level, and then work on the mismatch via an installer PR
3. merge the rest of this PR.

Does that still sound like the plan @jkyros

**Comment by jkyros** [comment]
- So #3286 merged
-  (and Colin had to roll back the extensions with #3290, whoops)
-  I've put up https://github.com/openshift/installer/pull/6234, which should make the machineconfigs match between bootstrap and live,
- Once that's done, we can come back here for the rest of the "use new format by default"

File: pkg/daemon/rpm-ostree.go [hunk 1]
```diff
@@ -306,6 +316,30 @@ func (r *RpmOstreeClient) Rebase(imgURL, osImageContentDir string) (changed bool
...
 
+// IsBootableImage determines if the image is a bootable (new container formet) image, rather than a wrapper (old container format)
+func (r *RpmOstreeClient) IsBootableImage(imgURL string) (bool, error) {
+
+	imageData, err := imageInspect(imgURL)
+	if err != nil {
+		return false, err
+	}
+
+	if isBootable, ok := imageData.Labels["ostree.bootable"]; ok {
```
Comment (cgwalters) [inline pkg/daemon/rpm-ostree.go:327]: I'm OK with this temporarily as a way to distinguish old vs new format, but I also think this check is better done in the ostree stack; done in https://github.com/ostreedev/ostree-rs-ext/pull/356

Once we drop the old `machine-os-content` and hard require new format, I don't think we need to duplicate the error we'd get from rpm-ostree anyways.

**Comment by cgwalters** [comment]
OK so this conflicts with https://github.com/openshift/machine-config-operator/pull/3302 right now - I think I vote we land this first (but disabled by default), then I'll rebase that PR and then rebase https://github.com/openshift/machine-config-operator/pull/3310 on top of that merged stack.

**Comment by cgwalters** [comment]
OK I took a crack at rebasing this myself, but man there are a ton of conflicts.  Is doing that on your near term radar?  If not I'll probably look at this later today

**Comment by jkyros** [comment]
Yes, I'll do that this afternoon. Most of this is already in in some fashion, what's left is really small. 

If we're going for shortest path to test "new format by default", I'm almost tempted to:
-  close this
-  hard code the extensions container to our CI extensions image for https://github.com/openshift/machine-config-operator/pull/3317 
- sneak in a DNS entry or IP address (or I guess extract the extensions to disk and use it that way).
- use that to test "new image by default", because with the extensions container that *should* at least have a chance of passing our CI

**Comment by jkyros** [comment]
Closing this in favor of https://github.com/openshift/machine-config-operator/pull/3317

**Pull request closed without merging**